bfa_rust/lib/heap.ml

```diff
@@ -0,0 +1,247 @@
+  SPmap.iter_vars_serialized
+    (Freeable.iter_vars_serialized Tree_block.iter_vars_serialized)
+    s
+
+let pp_pretty ~ignore_freed ft { heap; _ } =
+  let ignore =
+    if ignore_freed then function _, Freeable.Freed -> true | _ -> false
+    else fun _ -> false
+  in
+  match heap with
+  | None -> Fmt.pf ft "Empty Heap"
+  | Some st -> SPmap.pp ~ignore (Freeable.pp Tree_block.pp_pretty) ft st
+
+let empty = { heap = None; globals = GlobMap.empty }
+
+let log action ptr st =
+  L.debug (fun m ->
+      m "About to execute action: %s at %a (%a)@\n@[<2>HEAP:@ %a@]" action
+        Typed.ppa ptr Call_trace.pp_span (get_loc ())
+        (pp_pretty ~ignore_freed:true)
+        st)
+
+let with_ptr (ptr : [< T.sptr ] Typed.t) ({ heap; _ } as st : t)
+    (f :
+      ofs:[< T.sint ] Typed.t ->
+      Tree_block.t option ->
+      ('a * Tree_block.t option, 'err, 'fix list) Result.t) :
+    ('a * t, 'err, serialized list) Result.t =
+  let loc, ofs = Typed.Ptr.decompose ptr in
+  let++ v, heap = (SPmap.wrap (Freeable.wrap (f ~ofs))) loc heap in
+  (v, { st with heap })
+
+let with_ptr_read_only (ptr : [< T.sptr ] Typed.t) ({ heap; _ } : t)
+    (f :
+      ofs:[< T.sint ] Typed.t ->
+      Tree_block.t option ->
+      ('a, 'err, Tree_block.serialized list) Result.t) :
+    ('a, 'err, serialized list) Result.t =
+  let loc, ofs = Typed.Ptr.decompose ptr in
+  (SPmap.wrap_read_only (Freeable.wrap_read_only (f ~ofs))) loc heap
+
+let load ptr ty st =
+  let@ () = with_error_loc_as_call_trace () in
+  let@ () = with_loc_err () in
+  log "load" ptr st;
+  if%sat Typed.Ptr.is_at_null_loc ptr then Result.error `NullDereference
+  else
+    with_ptr ptr st (fun ~ofs block ->
+        L.debug (fun f ->
+            f "Recursively reading from block tree at %a:@.%a" Typed.ppa ptr
+              Fmt.(option ~none:(any "None") Tree_block.pp)
+              block);
+        let rec aux block = function
+          | `Done v -> Result.ok (v, block)
+          | `More (blocks, callback) ->
+              let pp_block ft (ty, size, ofs) =
+                Fmt.pf ft "%s [%a, +%a[ "
+                  (Charon_util.lit_to_string ty)
+                  Typed.ppa ofs Typed.ppa size
+              in
+              L.debug (fun f ->
+                  f "Loading blocks [%a]" Fmt.(list ~sep:comma pp_block) blocks);
+              let** values, block =
+                Result.fold_list blocks ~init:([], block)
+                  ~f:(fun (vals, block) (ty, size, ofs) ->
+                    let++ value, block = Tree_block.load ofs size ty block in
+                    (value :: vals, block))
+              in
+              let values = List.rev values in
+              let* res = callback values in
+              aux block res
+        in
+        let++ value, block = aux block (Layout.rust_of_cvals ~offset:ofs ty) in
+        L.debug (fun f ->
+            f "Finished reading rust value %a" Charon_util.pp_rust_val value);
+        (value, block))
```

Comment:
Ok so, for now this works.
It feels a tiny bit overkill but who cares.
I think that, anyway, when we get to more abstraction preservation and we have symbolic values denoting aggregates, this will probably go away. Because, if you don't exactly know the variant (like concretely), you can just put the whole thing into a symbolic value or something?